### PR TITLE
44635 categoty controller

### PR DESCRIPTION
**Pull request by Rownaknira**

### CODE REVIEW

File: app/assets/javascripts/categories.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (chuanhtuan) [inline app/assets/javascripts/categories.coffee]: remove unused file

File: app/assets/stylesheets/categories.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Categories controller here.
```
Comment (chuanhtuan) [inline app/assets/stylesheets/categories.scss]: remove unused file

File: app/controllers/categories_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Category.all
+  end
+  def new
+    @category	=	Category.new
```
Comment (chuanhtuan) [inline app/controllers/categories_controller.rb]: space

File: app/controllers/categories_controller.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  end
+  def show
+    @category	=	Category.find params[:id]
+  end
+  def create
+    @category	=	Category.new category_params
+		if @category.save
+			redirect_to :action => 'index'
```
Comment (chuanhtuan) [inline app/controllers/categories_controller.rb]: redirect_to categories_path

File: app/controllers/categories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		else
+			render 'new'
+		end
+  end
+  def edit
+    @category = Category.find params[:id]
+  end
+  def	update
+		@category	=	Category.find params[:id]
```
Comment (chuanhtuan) [inline app/controllers/categories_controller.rb]: indent

File: app/controllers/categories_controller.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+		if @category.update_attributes(category_params)
+			redirect_to :action => 'index'
```
Comment (chuanhtuan) [inline app/controllers/categories_controller.rb]: redirect_to categories_path

File: app/views/categories/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  <h1>Update Category</h1>
+  <br>
+    <div class="col-md-6 col-md-offset-3">
+	  <%= form_for(@category)	do |f| %>
+	    <div class="form-group">
+		  <%= f.label :name %>
+		  <%= f.text_field :name, class: 'form-control' %>
+		</div>
+		<br>
+		<center><%=	f.submit "Save changes", class: "btn btn-primary" %></center>
```
Comment (chuanhtuan) [inline app/views/categories/edit.html.erb]: i18n

File: app/views/categories/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<% provide(:title,	"Edit Category")	%>
+<div class="row">
+  <h1>Update Category</h1>
+  <br>
+    <div class="col-md-6 col-md-offset-3">
+	  <%= form_for(@category)	do |f| %>
```
Comment (chuanhtuan) [inline app/views/categories/edit.html.erb]: ()

File: app/views/categories/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    <tr>
+      <td><%= category.name%></td>
+      <td><%= link_to 'Edit', :action => 'edit', :id => category.id %></td>
+      <td><%= link_to 'delete',category,method: :delete,data:{ confirm: "Are you sure you want to delete?"}%>
+    </tr>
+   
+  <% end %>
+  </table>
+  <br>
+  <%= link_to 'New Category', new_category_path , class: "btn btn-primary" %>
```
Comment (chuanhtuan) [inline app/views/categories/index.html.erb]: i18n

File: app/views/categories/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  <h1>Add Category</h1>
+  <br>
+    <div class="col-md-6 col-md-offset-3">
+	  <%= form_for(@category)	do |f| %>
+	    <div class="form-group">
+		  <%= f.label :name %>
+		  <%= f.text_field :name, class: 'form-control' %>
+		</div>
+		<br>
+		<center><%=	f.submit "Create new Category", class: "btn btn-primary" %></center>
```
Comment (chuanhtuan) [inline app/views/categories/new.html.erb]: i18n

File: app/views/categories/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+<% provide(:title,	"New Category")	%>
+<div class="row">
+  <h1>Add Category</h1>
+  <br>
+    <div class="col-md-6 col-md-offset-3">
+	  <%= form_for(@category)	do |f| %>
```
Comment (chuanhtuan) [inline app/views/categories/new.html.erb]: ()

File: app/views/categories/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+<%= link_to 'Go Back', :back , class: "btn btn-primary"%>
+<%= link_to 'New Category', new_category_path , class: "btn btn-primary" %>
```
Comment (chuanhtuan) [inline app/views/categories/show.html.erb]: i18n

File: config/routes.rb [hunk 13]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
+  get 'categories/new'
+
   root	'static_pages#home'
+  
+  resources  :categories
```
Comment (chuanhtuan) [inline config/routes.rb]: use namespace admin

File: app/controllers/admin/categories_controller.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
+class Admin::CategoriesController < ApplicationController
+  def new
+    @category=Category.new
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: sapce
@category = Category.new

File: app/controllers/admin/categories_controller.rb [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+  end
+  def create
+    @category=Category.new category_params
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: space

File: app/controllers/admin/categories_controller.rb [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+		if @category.save
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: indent

File: app/controllers/admin/categories_controller.rb [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+			redirect_to categories_path
+		else
+			render 'new'
+		end
+  end
+  def edit
+    @category=Category.find params[:id]
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: space

File: app/controllers/admin/categories_controller.rb [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+  end
+  def	update
+		@category=Category.find params[:id]
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: indent

File: app/controllers/admin/categories_controller.rb [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+		if @category.update_attributes(category_params)
+			redirect_to categories_path
+		else
+			render 'edit'
+		end
+  end
+  def	destroy
+	  Category.find(params[:id]).destroy
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: indent

File: app/controllers/categories_controller.rb [hunk 20]
```diff
@@ -0,0 +1,8 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories=Category.all
```
Comment (chuanhtuan) [inline app/controllers/categories_controller.rb]: space

File: app/controllers/categories_controller.rb [hunk 21]
```diff
@@ -0,0 +1,8 @@
...
+  end
+  def show
+    @category=Category.find params[:id]
```
Comment (chuanhtuan) [inline app/controllers/categories_controller.rb]: space

File: app/views/admin/categories/edit.html.erb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+<% provide(:title,"Edit Category")%>
+<div class="row">
+  <h1>Update Category</h1>
```
Comment (chuanhtuan) [inline app/views/admin/categories/edit.html.erb]: i18n

File: app/helpers/categories_helper.rb [hunk 23]
```diff
@@ -0,0 +1,2 @@
+module CategoriesHelper
```
Comment (chuanhtuan) [inline app/helpers/categories_helper.rb]: remove unused file

File: app/views/admin/categories/edit.html.erb [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+  <br>
+    <div class="col-md-6 col-md-offset-3">
+	   <%=form_for [:admin, @category] do |f|%>
+	    <div class="form-group">
+		  <%= f.label :name %>
+		  <%= f.text_field :name, class: 'form-control' %>
+		</div>
+		<br>
+		<center><%=	f.submit t("form.update"), class: "btn btn-primary" %></center>
```
Comment (chuanhtuan) [inline app/views/admin/categories/edit.html.erb]: space

File: app/views/admin/categories/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,15 @@
+<% provide(:title,"New Category")	%>
+<div class="row">
+  <h1>Add Category</h1>
```
Comment (chuanhtuan) [inline app/views/admin/categories/new.html.erb]: i18n

File: app/views/categories/index.html.erb [hunk 26]
```diff
@@ -0,0 +1,21 @@
+<div class="row">
+  <h1>Listing of Categories</h1>
```
Comment (chuanhtuan) [inline app/views/categories/index.html.erb]: i18n

File: app/views/categories/index.html.erb [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+      <tr>
+        <th>Name</th>
+        <th colspan="2"></th>
+      </tr>
+    </thead>
+  <% @categories.each do |category| %>
+    <tr>
+      <td><%= category.name%></td>
+      <td><%= link_to t('form.edit'),edit_admin_category_path(category) %></td>
+      <td><%= link_to t('form.delete'),admin_category_path(category),method: :delete,data:{ confirm: "Are you sure you want to delete?"}%></td>
```
Comment (chuanhtuan) [inline app/views/categories/index.html.erb]: i18n

File: app/views/categories/show.html.erb [hunk 28]
```diff
@@ -0,0 +1,11 @@
+<%=link_to t('history.back'),:back,class: "btn btn-primary"%>
+<%=link_to t('form.create'),new_admin_category_path,class:"btn btn-primary"%>
+<br>
+<div class="row">
+  <h1>Category</h1>
```
Comment (chuanhtuan) [inline app/views/categories/show.html.erb]: i18n

File: app/controllers/admin/categories_controller.rb [hunk 29]
```diff
@@ -0,0 +1,33 @@
+class Admin::CategoriesController < ApplicationController
+	def new
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: indent 2

File: app/controllers/admin/categories_controller.rb [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+	    else
+	      render 'new'
+	    end
+	end
+	def edit
+	  @category = Category.find params[:id]
+	end
+	def	update
+		@category = Category.find params[:id]
+		  if @category.update_attributes(category_params)
```
Comment (chuanhtuan) [inline app/controllers/admin/categories_controller.rb]: ()

File: config/routes.rb [hunk 31]
```diff
@@ -5,4 +5,11 @@
   delete	'logout'  =>  'sessions#destroy'
   get	'signup'  =>  'users#new'
   resources	:users
-end
+
+  get 'categories/new'
```
Comment (chuanhtuan) [inline config/routes.rb]: remove

File: app/views/admin/categories/edit.html.erb [hunk 32]
```diff
@@ -0,0 +1,15 @@
+<% provide(:title,"Update Category")%>
+<div class="row">
+  <h1><%= t('category.edit')%></h1>
+  <br>
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for [:admin,@category] do |f|%>
```
Comment (chuanhtuan) [inline app/views/admin/categories/edit.html.erb]: space
<%= form_for [:admin, @category] do |f|%>

File: app/views/admin/categories/new.html.erb [hunk 33]
```diff
@@ -0,0 +1,15 @@
+<% provide(:title,"New Category")%>
+<div class="row">
+  <h1><%= t('category.new')%></h1>
+  <br>
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for [:admin,@category] do |f|%>
```
Comment (chuanhtuan) [inline app/views/admin/categories/new.html.erb]: space
<%= form_for [:admin, @category] do |f|%>

File: app/views/layouts/_header.html.erb [hunk 34]
```diff
@@ -1,16 +1,23 @@
...
-          <% if logged_in? %>
+    <nav>
+      <ul class="nav navbar-nav	navbar-right">
+        <% if logged_in? %>
           <li><%= link_to t('nav.home'),  root_path %></li>
           <li><%= link_to t('nav.users'), users_path %></li>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown" data-hover="dropdown" data-delay="1000" data-close-others="false">Category <b class="caret"></b></a>
+            <ul>
+              <li><%= link_to "Category List",categories_path,:tabindex => "-1"%></li>
```
Comment (chuanhtuan) [inline app/views/layouts/_header.html.erb]: <li><%= link_to "Category List", categories_path, tabindex: "-1"%></li>

File: app/views/layouts/_header.html.erb [hunk 35]
```diff
@@ -1,16 +1,23 @@
...
+              <li><%= link_to "New Category",new_admin_category_path,:tabindex => "-1"%></li>
```
Comment (chuanhtuan) [inline app/views/layouts/_header.html.erb]: <li><%= link_to "New Category", new_admin_category_path, tabindex: "-1"%></li>

File: config/routes.rb [hunk 36]
```diff
@@ -1,8 +1,15 @@
...
   root 'static_pages#home'
   get	 'login'  =>  'sessions#new'
   post  'login'  =>  'sessions#create'
   delete	'logout'  =>  'sessions#destroy'
   get	'signup'  =>  'users#new'
   resources	:users
+
+  resources :categories,only: [:index]
+  namespace :admin do
+    resources :categories,only: [:new,:create,:edit,:update,:destroy]
```
Comment (chuanhtuan) [inline config/routes.rb]: space
resources :categories, only: [:new,:create,:edit,:update,:destroy]

File: app/controllers/admin/categories_controller.rb [hunk 37]
```diff
@@ -0,0 +1,32 @@
+class Admin::CategoriesController < ApplicationController
+  def new
+    @category = Category.new	
+  end
+  def create
```
Comment (chikim) [inline app/controllers/admin/categories_controller.rb]: should have an empty line between 2 functions declaration

File: app/controllers/admin/categories_controller.rb [hunk 38]
```diff
@@ -0,0 +1,32 @@
...
+    @category = Category.new category_params
+    if @category.save
+      redirect_to categories_path
+    else
+      render 'new'
+    end    
+  end
+  def edit
```
Comment (chikim) [inline app/controllers/admin/categories_controller.rb]: should have an empty line between 2 functions declaration

File: app/controllers/admin/categories_controller.rb [hunk 39]
```diff
@@ -0,0 +1,32 @@
...
+    @category = Category.find params[:id]
+  end
+  def update
```
Comment (chikim) [inline app/controllers/admin/categories_controller.rb]: should have an empty line between 2 functions declaration

File: app/controllers/admin/categories_controller.rb [hunk 40]
```diff
@@ -0,0 +1,32 @@
...
+    @category = Category.find params[:id]
+    if @category.update_attributes category_params
+      redirect_to categories_path
+    else
+      render 'edit'
+    end    
+  end
+  def destroy
```
Comment (chikim) [inline app/controllers/admin/categories_controller.rb]: should have an empty line between 2 functions declaration

File: app/controllers/admin/categories_controller.rb [hunk 41]
```diff
@@ -0,0 +1,32 @@
...
+    Category.find(params[:id]).destroy
+    redirect_to categories_url
+  end
+  private
```
Comment (chikim) [inline app/controllers/admin/categories_controller.rb]: should have an empty line between 2 functions declaration

File: app/controllers/admin/categories_controller.rb [hunk 42]
```diff
@@ -0,0 +1,32 @@
...
+    def category_params
```
Comment (chikim) [inline app/controllers/admin/categories_controller.rb]: private functions should be indented by the same column with `private`

File: app/views/admin/categories/edit.html.erb [hunk 43]
```diff
@@ -0,0 +1,15 @@
...
+  <br>
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for [:admin, @category] do |f|%>
+      <div class="form-group">
+        <%= f.label :name%>
+        <%= f.text_field :name,class: 'form-control'%>
+      </div>
+      <br>
+      <%= f.submit t("form.update"),class: "btn btn-primary"%>	
+      <% end %>
```
Comment (chikim) [inline app/views/admin/categories/edit.html.erb]: wrong indent

File: app/views/admin/categories/new.html.erb [hunk 44]
```diff
@@ -0,0 +1,15 @@
...
+  <br>
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for [:admin, @category] do |f|%>
+      <div class="form-group">
+        <%= f.label :name%>
+        <%= f.text_field :name,class: 'form-control'%>
+      </div>
+      <br>
+      <%= f.submit t("form.create"),class: "btn btn-primary"%>	
+      <% end %>
```
Comment (chikim) [inline app/views/admin/categories/new.html.erb]: wrong indent

File: app/views/categories/index.html.erb [hunk 45]
```diff
@@ -0,0 +1,21 @@
...
+    <% @categories.each do |category|%>
+      <tr>
+        <td><%= category.name%></td>
+        <td><%= link_to t('form.edit'),edit_admin_category_path(category)%></td>
+        <td><%= link_to t('form.delete'),admin_category_path(category),method: :delete,data: {confirm: t('delete.confirmation')}%></td>
+      </tr>
+    <% end %>	
+  </table>
+  <br>
+    <%= link_to t('form.create'),new_admin_category_path,class: "btn btn-primary"%>
```
Comment (chikim) [inline app/views/categories/index.html.erb]: wrong indent

File: app/views/layouts/_header.html.erb [hunk 46]
```diff
@@ -1,16 +1,23 @@
-<header	class="navbar	navbar-fixed-top	navbar-inverse">
-  <div	class="container">
-      <nav>
-      <ul	class="nav navbar-nav	navbar-right">
-          <% if logged_in? %>
+<header class="navbar navbar-fixed-top navbar-inverse">
+  <div class="container">
+    <nav>
+      <ul class="nav navbar-nav	navbar-right">
```
Comment (chikim) [inline app/views/layouts/_header.html.erb]: spaces before close quotation mark

File: config/database.yml.example [hunk 47]
```diff
@@ -0,0 +1,25 @@
+# SQLite version 3.x
+#   gem install sqlite3
+#
+#   Ensure the SQLite 3 gem is defined in your Gemfile
+#   gem 'sqlite3'
+#
+default: &default
+  adapter: sqlite3
```
Comment (chikim) [inline config/database.yml.example]: why don't we use mysql

File: config/routes.rb [hunk 48]
```diff
@@ -1,8 +1,15 @@
 Rails.application.routes.draw do
+  get 'categories/new'
```
Comment (chikim) [inline config/routes.rb]: is it neccesary?

**Pull request merged**